### ISSUE TITLE
investigate "recover" feature

### ISSUE THREAD

**Issue by Geal** [Geal/nom#499]
it is the ability to recognize an error and register it but continue parsing
https://github.com/jaspervdj/talks/blob/master/2017-skillsmatter-errors/slides.md#on-recovery-2

**Comment by lukechu10** [comment]
@ebkalderon thanks for your detailed and informative blog post. I was wandering how I could implement the error recovery strategy described in your blog. A problem I can see is that if I use an `alt` combinator, only the first parser will be called because it will always succeed. How would I fix this problem? Thanks!

**Comment by ebkalderon** [comment]
@lukechu10 The point of the described technique is for the _top-level_ parser to always succeed, but not necessarily all of its component parts. Every possible failure case must be handled such that the top-level parser always produces some result, even if said result is partial and is accompanied by a stack of parse errors, and the post describes one possible method for expressing this behavior ergonomically. If you read the blog post, you will see that fallible parser combinators, including the `alt()` combinator itself, are still being used internally.

**Comment by lukechu10** [comment]
> @lukechu10 The point of the described technique is for the _top-level_ parser to always succeed, but not necessarily all of its component parts. Every possible failure case must be handled such that the top-level parser always produces some result, even if said result is partial and is accompanied by a stack of parse errors, and the post describes one possible method for expressing this behavior ergonomically. If you read the blog post, you will see that fallible parser combinators, including the `alt()` combinator itself, are still being used internally.

@ebkalderon If I understand correctly, `expect` can only be used if no lookahead is needed. Correct? Otherwise, the other `alt` branches would never get executed. Say for example we had the following grammar:
```ebnf
root := paren | empty_paren
paren := '(' expr ')'
empty_paren := '()'
expr := a number literal
```
In a non recovering scenario, the parser for `root` would be something like:
```rust
alt((paren, empty_paren))
```
Parsing the string `()` would fail on `paren` and than try to parse it using `empty_paren`.

In an error recovering scenario, it would emit the errors:
```
unexpected `)`, expected an expression
```
instead of parsing an `empty_paren`.

**Comment by ebkalderon** [comment]
@lukechu10 Yep, that's correct! The `expect` combinator is intended to be applied when a specific branch of `alt()` is determined to be correct, but contains recoverable (that is, non-fatal) syntax errors. The point is to use `alt()` with lookahead to parse the minimum required information to determine the correct branch to use, and once it's known, use `expect()` for the rest to recover gracefully from errors.

For example, a parser for an expression-oriented programming language like [Nix](https://nixos.wiki/wiki/Nix_Expression_Language) could leverage `alt()` to determine whether a sequence beginning with `{` is a set or a function. Once parsing has advanced far enough that we are confident the current branch is correct, `expect()` can be used to parse the contents of the set or function body and recover gracefully from syntax errors, e.g. marking functions with invalid bodies as `Expr::Error` and pushing some error messages onto the error stack while allowing parsing to skip ahead and continue.
